Dataanalyse og Machine Learning (F24.520214U002.A)  
Gruppe 1:  
Hans Christian Hulvej - AU19930532  
Kathrine Bohus Madsen - AU202301227  
Tor Kolding - AU20000148  
Torben Holst Rendboe – AU20064217    
  
Opgave 3  
  
Deep Learning classification tensorflow med tensorboard  

Iterativ gennemløb af: *epochs*, *neurons*, *layers*, *dropout* & *Optimazion type* holdt op mod *accuracy* med henblik på at finde den **optimale kombination**.

D. 26-04-2024

In [367]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
import datetime
import itertools
import os
import shutil

# Load the TensorBoard notebook extension.
%reload_ext tensorboard

Vi starter med at slette tidligere tensorboard logs.  
Hvis man vil vise alle logs, udkommenteres / slettes disse to liner kode.

In [368]:
shutil.rmtree('logs/fit') 
os.mkdir('logs/fit')

Vi starter med at hente datasettet, i dette tilfælde er det **WineQT.csv**  
Så fjerne vi Id kolonnen, da den ikke tjener noget formål og skriver lige info ud for at tjekke at det er som vi ønsker.  
Så deler vi datasettet op i *features* og *target*. 

In [369]:
data = pd.read_csv('WineQT.csv')
data = data.drop(columns=['Id'])
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1143 non-null   float64
 1   volatile acidity      1143 non-null   float64
 2   citric acid           1143 non-null   float64
 3   residual sugar        1143 non-null   float64
 4   chlorides             1143 non-null   float64
 5   free sulfur dioxide   1143 non-null   float64
 6   total sulfur dioxide  1143 non-null   float64
 7   density               1143 non-null   float64
 8   pH                    1143 non-null   float64
 9   sulphates             1143 non-null   float64
 10  alcohol               1143 non-null   float64
 11  quality               1143 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 107.3 KB


In [370]:
features = data.drop(columns=["quality"])
target = data["quality"]

Vi anvender **StandarScaler** klassen til at normaliserer vores *features*, og gemmer resultatet i *X_scaled*.  
Vi gemmer også target i *y*  
  
Så deler vi vores data (*X_scaled* og *y*) op i train-set og test-set, med en fordeling på 80% til træning og 20% til testning.  
Vi sætter og *random_state* til 42 (fast seed).

In [371]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(features)
y = target

In [372]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

<div class="alert-warning" style="border-style: solid; padding:5px;">
<h2>FAKTA BOKS</h2>  
    
I denne model, arbejder vi med keras sequential model, som i princippet er en stak af lag, men især kendetegnet ved at hvert lag har en input- og en outputtensor. I dette tilfælde er der dog indtil videre kun et lag.  
  
Det er her i modellen at vi definerer hvormange fuldt forbundene neuroner der skal være (inden dropout).  
  
Derefter vælger vi hvor stort vores dropout skal være.
</div>

Vi bygger vores sequential model med et lag hvoi der er 128 neuroner, med **relu** som aktiveringstype og et dropout på 20 %

In [373]:
model = tf.keras.Sequential([
    layers.Dense(128, activation='relu', input_shape=(11,)),
    layers.Dropout(0.2),
    layers.Dense(11)  # Output layer with 11 units (one for each quality score)
])

<div class="alert-warning" style="border-style: solid; padding:5px;">
<h2>FAKTA BOKS</h2>  
  
Så skal vi have modellen konfigureret til træning, det gør vi ved at *compilerer* den.  
  
Når vi compilerer, skal vi vælge hvilken optimeringsalgoritme vi ønsker at bruge, i dette tilfælde er det **adam**, som bestemmer måden hvorpå modellen evaluerer.  
  
Og tilsidt konfigurer vi de/den data vi ønsker at overvåge under testen.  

</div>

Vi compilerer modellen med **adam** som optmeringsmetode og accuracy som den parameter vi vil se på.

In [374]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

<div class="alert-warning" style="border-style: solid; padding:5px;">
<h2>FAKTA BOKS</h2> 

Nu skal modellen trænes, med **model.fit**.  
  
Vi konfigurer *træningsdata* og *valideringsdata*.
  
Her konfigurer vi også hvormange gange vores træningssæt bliver "kørt" gennem modellen, *epochs*
    
Sluttelig sørger vi for at data bliver gjort tilgængelig for tensorboard, ved hjælp at *callbacks*.  
  
</div>

Så træner vi modellen med vores trænings data og valideringsdata, med en gennemkørsel på 20 gange.  
  
Inden da konfigurer vi stien til den placering hvor tensorboard skal hente sine logfiler og konfigurer vores callbacks (tensorboards og earlystop).

In [375]:
logdir="logs/fit/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = keras.callbacks.TensorBoard(log_dir=logdir) # eller "squared error" ved regression
early_stopping_callback = keras.callbacks.EarlyStopping(monitor='val_loss', patience=3)

model.fit(X_train, y_train, epochs=20, validation_data=(X_test, y_test), callbacks=[tensorboard_callback, early_stopping_callback]) # 

Epoch 1/20
29/29 [==============================] - 1s 12ms/step - loss: 2.0940 - accuracy: 0.2976 - val_loss: 1.6536 - val_accuracy: 0.5415
Epoch 2/20
29/29 [==============================] - 0s 5ms/step - loss: 1.5013 - accuracy: 0.5481 - val_loss: 1.2519 - val_accuracy: 0.5764
Epoch 3/20
29/29 [==============================] - 0s 6ms/step - loss: 1.2426 - accuracy: 0.5427 - val_loss: 1.0808 - val_accuracy: 0.5808
Epoch 4/20
29/29 [==============================] - 0s 6ms/step - loss: 1.1382 - accuracy: 0.5667 - val_loss: 1.0065 - val_accuracy: 0.6026
Epoch 5/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0714 - accuracy: 0.5799 - val_loss: 0.9679 - val_accuracy: 0.6201
Epoch 6/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0412 - accuracy: 0.5919 - val_loss: 0.9425 - val_accuracy: 0.6245
Epoch 7/20
29/29 [==============================] - 0s 6ms/step - loss: 1.0287 - accuracy: 0.5853 - val_loss: 0.9268 - val_accuracy: 0.6201
Epoch 8/20
29/29 [=

Vi starter tensorboard  
  
Note: Hvis tensorboard ikke vil starte kan det skyldes at der ligger en PID fil og blokkere. Den kan fjernes ved at gå ind i  
c:\\bruger\<dit brugerid>\AppData\AppData\Local\Temp\.tensorboard-info og slette de PID filer der ligger der.

In [376]:
%tensorboard --logdir logs

Reusing TensorBoard on port 6006 (pid 23980), started 23:26:19 ago. (Use '!kill 23980' to kill it.)

In [377]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print("Test accuracy:", "%.2f" % (test_acc*100), " %")

8/8 [==============================] - 0s 3ms/step - loss: 0.8818 - accuracy: 0.6201
Test accuracy: 62.01  %


### Gennemløb med ændrede parameter ###
  
Så laver vi en optimeringstest, hvor vi iterer over epochs, neuron, dropout og optimerings metode.

In [378]:
no_epochs = [5, 10, 15, 20]
no_neurons = [64, 128, 256]
dropouts = [0, 0.1, 0.2, 0.3, 0.4, 0.5]
optim_types = ['adam', 'sgd', 'rmsprop']
combs = list(itertools.product(no_epochs, no_neurons, dropouts, optim_types))

In [379]:
test_results = pd.DataFrame(columns=['epochs', 'neurons', 'dropout', 'Optimazion type', 'accuracy'])
test_results = test_results.head(0)

In [380]:
start_time = datetime.datetime.now()

In [381]:
for no_epochs, no_neurons, dropouts, optim_types in combs:
    model = tf.keras.Sequential([
        layers.Dense(no_neurons, activation='relu', input_shape=(11,)),
        layers.Dense(no_neurons, activation='relu'),
        layers.Dropout(dropouts),
        layers.Dense(11)  # Output layer with 11 units (one for each quality score)
    ])
    model.compile(optimizer=optim_types,
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
    model.fit(X_train, y_train, epochs=no_epochs, validation_data=(X_test, y_test))
    acc = model.evaluate(X_test, y_test)[1]
    params = {'epochs':no_epochs, 'neurons':no_neurons, 'dropout': dropouts, 'Optimazion type': optim_types, 'accuracy': acc}
    test_results.loc[len(test_results)] = params

Epoch 1/5
29/29 [==============================] - 2s 16ms/step - loss: 1.8377 - accuracy: 0.4114 - val_loss: 1.3034 - val_accuracy: 0.5808
Epoch 2/5
29/29 [==============================] - 0s 5ms/step - loss: 1.2435 - accuracy: 0.5492 - val_loss: 1.0193 - val_accuracy: 0.5939
Epoch 3/5
29/29 [==============================] - 0s 5ms/step - loss: 1.0842 - accuracy: 0.5755 - val_loss: 0.9510 - val_accuracy: 0.6419
Epoch 4/5
29/29 [==============================] - 0s 5ms/step - loss: 1.0247 - accuracy: 0.5952 - val_loss: 0.9223 - val_accuracy: 0.6463
Epoch 5/5
8/8 [==============================] - 0s 3ms/step - loss: 0.9008 - accuracy: 0.6332
Epoch 1/5
29/29 [==============================] - 1s 14ms/step - loss: 2.3128 - accuracy: 0.1937 - val_loss: 2.1847 - val_accuracy: 0.3362
Epoch 2/5
29/29 [==============================] - 0s 5ms/step - loss: 2.0625 - accuracy: 0.4354 - val_loss: 1.9372 - val_accuracy: 0.4847
Epoch 3/5
29/29 [==============================] - 0s 5ms/step - loss

Epoch 4/5
29/29 [==============================] - 0s 6ms/step - loss: 1.5393 - accuracy: 0.4912 - val_loss: 1.3638 - val_accuracy: 0.5109
Epoch 5/5
8/8 [==============================] - 0s 3ms/step - loss: 1.2674 - accuracy: 0.5371
Epoch 1/5
29/29 [==============================] - 2s 18ms/step - loss: 1.7189 - accuracy: 0.4497 - val_loss: 1.2525 - val_accuracy: 0.5808
Epoch 2/5
29/29 [==============================] - 0s 10ms/step - loss: 1.2613 - accuracy: 0.5602 - val_loss: 1.0441 - val_accuracy: 0.6114
Epoch 3/5
29/29 [==============================] - 0s 7ms/step - loss: 1.1649 - accuracy: 0.5635 - val_loss: 0.9799 - val_accuracy: 0.6157
Epoch 4/5
29/29 [==============================] - 0s 5ms/step - loss: 1.0823 - accuracy: 0.5919 - val_loss: 0.9416 - val_accuracy: 0.6419
Epoch 5/5
8/8 [==============================] - 0s 4ms/step - loss: 0.9234 - accuracy: 0.6419
Epoch 1/5
29/29 [==============================] - 1s 12ms/step - loss: 1.8604 - accuracy: 0.3479 - val_loss: 1.3

29/29 [==============================] - 1s 14ms/step - loss: 1.6402 - accuracy: 0.4759 - val_loss: 1.1243 - val_accuracy: 0.5764
Epoch 2/5
29/29 [==============================] - 0s 5ms/step - loss: 1.1055 - accuracy: 0.5711 - val_loss: 0.9693 - val_accuracy: 0.6332
Epoch 3/5
29/29 [==============================] - 0s 5ms/step - loss: 1.0272 - accuracy: 0.5941 - val_loss: 0.9365 - val_accuracy: 0.6157
Epoch 4/5
29/29 [==============================] - 0s 5ms/step - loss: 0.9783 - accuracy: 0.6083 - val_loss: 0.9043 - val_accuracy: 0.6419
Epoch 5/5
8/8 [==============================] - 0s 4ms/step - loss: 0.8927 - accuracy: 0.6332
Epoch 1/5
29/29 [==============================] - 1s 18ms/step - loss: 2.2492 - accuracy: 0.2768 - val_loss: 2.0536 - val_accuracy: 0.5284
Epoch 2/5
29/29 [==============================] - 0s 7ms/step - loss: 1.9145 - accuracy: 0.4934 - val_loss: 1.7407 - val_accuracy: 0.5328
Epoch 3/5
29/29 [==============================] - 0s 6ms/step - loss: 1.6565 -

Epoch 4/5
29/29 [==============================] - 0s 4ms/step - loss: 1.4634 - accuracy: 0.5284 - val_loss: 1.3081 - val_accuracy: 0.5590
Epoch 5/5
8/8 [==============================] - 0s 2ms/step - loss: 1.2174 - accuracy: 0.5721
Epoch 1/5
29/29 [==============================] - 1s 12ms/step - loss: 1.5916 - accuracy: 0.4902 - val_loss: 1.1114 - val_accuracy: 0.5721
Epoch 2/5
29/29 [==============================] - 0s 4ms/step - loss: 1.1468 - accuracy: 0.5503 - val_loss: 0.9900 - val_accuracy: 0.5983
Epoch 3/5
29/29 [==============================] - 0s 4ms/step - loss: 1.0450 - accuracy: 0.5941 - val_loss: 0.9293 - val_accuracy: 0.6288
Epoch 4/5
29/29 [==============================] - 0s 4ms/step - loss: 1.0109 - accuracy: 0.5941 - val_loss: 0.9138 - val_accuracy: 0.6157
Epoch 5/5
8/8 [==============================] - 0s 2ms/step - loss: 0.8904 - accuracy: 0.6594
Epoch 1/5
29/29 [==============================] - 2s 16ms/step - loss: 1.7971 - accuracy: 0.4234 - val_loss: 1.17

29/29 [==============================] - 1s 11ms/step - loss: 1.5366 - accuracy: 0.5044 - val_loss: 1.0247 - val_accuracy: 0.6026
Epoch 2/5
29/29 [==============================] - 0s 5ms/step - loss: 1.0508 - accuracy: 0.6149 - val_loss: 0.9382 - val_accuracy: 0.6288
Epoch 3/5
29/29 [==============================] - 0s 5ms/step - loss: 1.0007 - accuracy: 0.6039 - val_loss: 0.9508 - val_accuracy: 0.5764
Epoch 4/5
29/29 [==============================] - 0s 5ms/step - loss: 0.9539 - accuracy: 0.6171 - val_loss: 0.8966 - val_accuracy: 0.6332
Epoch 5/5
8/8 [==============================] - 0s 2ms/step - loss: 0.9063 - accuracy: 0.6114
Epoch 1/5
29/29 [==============================] - 1s 13ms/step - loss: 2.3734 - accuracy: 0.1641 - val_loss: 2.1461 - val_accuracy: 0.4279
Epoch 2/5
29/29 [==============================] - 0s 4ms/step - loss: 1.9933 - accuracy: 0.4606 - val_loss: 1.7905 - val_accuracy: 0.4978
Epoch 3/5
29/29 [==============================] - 0s 5ms/step - loss: 1.6983 -

Epoch 4/5
29/29 [==============================] - 0s 5ms/step - loss: 1.4425 - accuracy: 0.5317 - val_loss: 1.2791 - val_accuracy: 0.5808
Epoch 5/5
8/8 [==============================] - 0s 2ms/step - loss: 1.1887 - accuracy: 0.5895
Epoch 1/5
29/29 [==============================] - 1s 13ms/step - loss: 1.3779 - accuracy: 0.5208 - val_loss: 0.9888 - val_accuracy: 0.6157
Epoch 2/5
29/29 [==============================] - 0s 5ms/step - loss: 1.0753 - accuracy: 0.5799 - val_loss: 0.9385 - val_accuracy: 0.6201
Epoch 3/5
29/29 [==============================] - 0s 5ms/step - loss: 1.0208 - accuracy: 0.5941 - val_loss: 0.9163 - val_accuracy: 0.6070
Epoch 4/5
29/29 [==============================] - 0s 6ms/step - loss: 0.9847 - accuracy: 0.6116 - val_loss: 0.9208 - val_accuracy: 0.5852
Epoch 5/5
8/8 [==============================] - 0s 2ms/step - loss: 0.9150 - accuracy: 0.5895
Epoch 1/10
29/29 [==============================] - 1s 11ms/step - loss: 1.9037 - accuracy: 0.4114 - val_loss: 1.4

Epoch 9/10
29/29 [==============================] - 0s 6ms/step - loss: 1.1627 - accuracy: 0.5678 - val_loss: 1.0503 - val_accuracy: 0.5983
Epoch 10/10
8/8 [==============================] - 0s 3ms/step - loss: 1.0292 - accuracy: 0.6114
Epoch 1/10
29/29 [==============================] - 1s 11ms/step - loss: 1.6245 - accuracy: 0.4945 - val_loss: 1.2175 - val_accuracy: 0.5415
Epoch 2/10
29/29 [==============================] - 0s 4ms/step - loss: 1.1973 - accuracy: 0.5602 - val_loss: 1.0186 - val_accuracy: 0.5895
Epoch 3/10
29/29 [==============================] - 0s 4ms/step - loss: 1.0795 - accuracy: 0.5799 - val_loss: 0.9352 - val_accuracy: 0.6026
Epoch 4/10
29/29 [==============================] - 0s 4ms/step - loss: 1.0291 - accuracy: 0.6083 - val_loss: 0.9211 - val_accuracy: 0.6245
Epoch 5/10
29/29 [==============================] - 0s 5ms/step - loss: 0.9892 - accuracy: 0.6094 - val_loss: 0.9030 - val_accuracy: 0.6245
Epoch 6/10
29/29 [==============================] - 0s 5ms/ste

Epoch 4/10
29/29 [==============================] - 0s 4ms/step - loss: 1.4446 - accuracy: 0.4825 - val_loss: 1.2962 - val_accuracy: 0.5415
Epoch 5/10
29/29 [==============================] - 0s 4ms/step - loss: 1.3711 - accuracy: 0.5230 - val_loss: 1.2169 - val_accuracy: 0.5677
Epoch 6/10
29/29 [==============================] - 0s 4ms/step - loss: 1.3000 - accuracy: 0.5394 - val_loss: 1.1590 - val_accuracy: 0.5590
Epoch 7/10
29/29 [==============================] - 0s 4ms/step - loss: 1.2717 - accuracy: 0.5449 - val_loss: 1.1135 - val_accuracy: 0.5764
Epoch 8/10
29/29 [==============================] - 0s 4ms/step - loss: 1.2212 - accuracy: 0.5460 - val_loss: 1.0779 - val_accuracy: 0.5852
Epoch 9/10
29/29 [==============================] - 0s 4ms/step - loss: 1.1835 - accuracy: 0.5689 - val_loss: 1.0505 - val_accuracy: 0.5852
Epoch 10/10
8/8 [==============================] - 0s 2ms/step - loss: 1.0295 - accuracy: 0.5852
Epoch 1/10
29/29 [==============================] - 1s 12ms/ste

29/29 [==============================] - 0s 4ms/step - loss: 1.0336 - accuracy: 0.5908 - val_loss: 0.9108 - val_accuracy: 0.6026
Epoch 10/10
8/8 [==============================] - 0s 2ms/step - loss: 0.8987 - accuracy: 0.6245
Epoch 1/10
29/29 [==============================] - 1s 12ms/step - loss: 2.2728 - accuracy: 0.2505 - val_loss: 1.9971 - val_accuracy: 0.4367
Epoch 2/10
29/29 [==============================] - 0s 4ms/step - loss: 1.9459 - accuracy: 0.4114 - val_loss: 1.6876 - val_accuracy: 0.5153
Epoch 3/10
29/29 [==============================] - 0s 4ms/step - loss: 1.6973 - accuracy: 0.4705 - val_loss: 1.4788 - val_accuracy: 0.5415
Epoch 4/10
29/29 [==============================] - 0s 4ms/step - loss: 1.5564 - accuracy: 0.4902 - val_loss: 1.3454 - val_accuracy: 0.5546
Epoch 5/10
29/29 [==============================] - 0s 4ms/step - loss: 1.4256 - accuracy: 0.5186 - val_loss: 1.2524 - val_accuracy: 0.5721
Epoch 6/10
29/29 [==============================] - 0s 4ms/step - loss: 1

Epoch 4/10
29/29 [==============================] - 0s 5ms/step - loss: 0.9889 - accuracy: 0.6182 - val_loss: 0.9090 - val_accuracy: 0.6463
Epoch 5/10
29/29 [==============================] - 0s 5ms/step - loss: 0.9471 - accuracy: 0.6236 - val_loss: 0.9128 - val_accuracy: 0.6157
Epoch 6/10
29/29 [==============================] - 0s 5ms/step - loss: 0.9278 - accuracy: 0.6258 - val_loss: 0.9080 - val_accuracy: 0.6201
Epoch 7/10
29/29 [==============================] - 0s 16ms/step - loss: 0.9108 - accuracy: 0.6313 - val_loss: 0.8877 - val_accuracy: 0.6201
Epoch 8/10
29/29 [==============================] - 0s 5ms/step - loss: 0.8916 - accuracy: 0.6466 - val_loss: 0.8886 - val_accuracy: 0.6070
Epoch 9/10
29/29 [==============================] - 0s 7ms/step - loss: 0.8723 - accuracy: 0.6510 - val_loss: 0.8933 - val_accuracy: 0.5983
Epoch 10/10
8/8 [==============================] - 0s 2ms/step - loss: 0.8866 - accuracy: 0.6114
Epoch 1/10
29/29 [==============================] - 1s 13ms/st

29/29 [==============================] - 0s 5ms/step - loss: 0.9047 - accuracy: 0.6302 - val_loss: 0.8947 - val_accuracy: 0.5895
Epoch 10/10
8/8 [==============================] - 0s 3ms/step - loss: 0.8832 - accuracy: 0.5983
Epoch 1/10
29/29 [==============================] - 2s 17ms/step - loss: 1.7164 - accuracy: 0.4661 - val_loss: 1.1078 - val_accuracy: 0.6070
Epoch 2/10
29/29 [==============================] - 0s 8ms/step - loss: 1.1442 - accuracy: 0.5689 - val_loss: 0.9591 - val_accuracy: 0.6245
Epoch 3/10
29/29 [==============================] - 0s 7ms/step - loss: 1.0494 - accuracy: 0.5864 - val_loss: 0.9248 - val_accuracy: 0.6201
Epoch 4/10
29/29 [==============================] - 0s 6ms/step - loss: 1.0165 - accuracy: 0.5985 - val_loss: 0.9265 - val_accuracy: 0.6114
Epoch 5/10
29/29 [==============================] - 0s 6ms/step - loss: 0.9831 - accuracy: 0.6171 - val_loss: 0.9043 - val_accuracy: 0.6288
Epoch 6/10
29/29 [==============================] - 0s 5ms/step - loss: 0

Epoch 4/10
29/29 [==============================] - 0s 7ms/step - loss: 1.0068 - accuracy: 0.5985 - val_loss: 0.9160 - val_accuracy: 0.6070
Epoch 5/10
29/29 [==============================] - 0s 6ms/step - loss: 0.9881 - accuracy: 0.5908 - val_loss: 0.8923 - val_accuracy: 0.6245
Epoch 6/10
29/29 [==============================] - 0s 6ms/step - loss: 0.9578 - accuracy: 0.6193 - val_loss: 0.8899 - val_accuracy: 0.5983
Epoch 7/10
29/29 [==============================] - 0s 6ms/step - loss: 0.9450 - accuracy: 0.6258 - val_loss: 0.8713 - val_accuracy: 0.6245
Epoch 8/10
29/29 [==============================] - 0s 6ms/step - loss: 0.9348 - accuracy: 0.6258 - val_loss: 0.8974 - val_accuracy: 0.6026
Epoch 9/10
29/29 [==============================] - 0s 7ms/step - loss: 0.9193 - accuracy: 0.6127 - val_loss: 0.8804 - val_accuracy: 0.6026
Epoch 10/10
8/8 [==============================] - 0s 3ms/step - loss: 0.8901 - accuracy: 0.5983
Epoch 1/10
29/29 [==============================] - 2s 14ms/ste

29/29 [==============================] - 0s 4ms/step - loss: 1.1066 - accuracy: 0.5821 - val_loss: 1.0091 - val_accuracy: 0.6026
Epoch 10/10
8/8 [==============================] - 0s 2ms/step - loss: 0.9886 - accuracy: 0.6201
Epoch 1/10
29/29 [==============================] - 1s 14ms/step - loss: 1.3295 - accuracy: 0.5339 - val_loss: 0.9669 - val_accuracy: 0.6245
Epoch 2/10
29/29 [==============================] - 0s 5ms/step - loss: 1.0026 - accuracy: 0.6083 - val_loss: 1.0013 - val_accuracy: 0.5459
Epoch 3/10
29/29 [==============================] - 0s 4ms/step - loss: 0.9588 - accuracy: 0.6083 - val_loss: 0.9023 - val_accuracy: 0.6288
Epoch 4/10
29/29 [==============================] - 0s 5ms/step - loss: 0.9258 - accuracy: 0.6357 - val_loss: 0.9100 - val_accuracy: 0.6245
Epoch 5/10
29/29 [==============================] - 0s 9ms/step - loss: 0.8929 - accuracy: 0.6510 - val_loss: 0.9289 - val_accuracy: 0.6026
Epoch 6/10
29/29 [==============================] - 0s 5ms/step - loss: 0

Epoch 4/10
29/29 [==============================] - 0s 5ms/step - loss: 1.3871 - accuracy: 0.5350 - val_loss: 1.2706 - val_accuracy: 0.5197
Epoch 5/10
29/29 [==============================] - 0s 5ms/step - loss: 1.2933 - accuracy: 0.5449 - val_loss: 1.1832 - val_accuracy: 0.5415
Epoch 6/10
29/29 [==============================] - 0s 6ms/step - loss: 1.2263 - accuracy: 0.5558 - val_loss: 1.1258 - val_accuracy: 0.5502
Epoch 7/10
29/29 [==============================] - 0s 17ms/step - loss: 1.1824 - accuracy: 0.5613 - val_loss: 1.0870 - val_accuracy: 0.5633
Epoch 8/10
29/29 [==============================] - 0s 5ms/step - loss: 1.1484 - accuracy: 0.5667 - val_loss: 1.0538 - val_accuracy: 0.5808
Epoch 9/10
29/29 [==============================] - 0s 6ms/step - loss: 1.1237 - accuracy: 0.5777 - val_loss: 1.0287 - val_accuracy: 0.5852
Epoch 10/10
8/8 [==============================] - 0s 7ms/step - loss: 1.0108 - accuracy: 0.5895
Epoch 1/10
29/29 [==============================] - 2s 25ms/st

29/29 [==============================] - 0s 7ms/step - loss: 0.8823 - accuracy: 0.6444 - val_loss: 0.9142 - val_accuracy: 0.6070
Epoch 10/10
8/8 [==============================] - 0s 4ms/step - loss: 0.8865 - accuracy: 0.6376
Epoch 1/10
29/29 [==============================] - 1s 14ms/step - loss: 2.2416 - accuracy: 0.2681 - val_loss: 1.9861 - val_accuracy: 0.5066
Epoch 2/10
29/29 [==============================] - 0s 5ms/step - loss: 1.8645 - accuracy: 0.4803 - val_loss: 1.6389 - val_accuracy: 0.5721
Epoch 3/10
29/29 [==============================] - 0s 6ms/step - loss: 1.5742 - accuracy: 0.5197 - val_loss: 1.4074 - val_accuracy: 0.5633
Epoch 4/10
29/29 [==============================] - 0s 7ms/step - loss: 1.4068 - accuracy: 0.5481 - val_loss: 1.2641 - val_accuracy: 0.5808
Epoch 5/10
29/29 [==============================] - 0s 7ms/step - loss: 1.2923 - accuracy: 0.5383 - val_loss: 1.1750 - val_accuracy: 0.5852
Epoch 6/10
29/29 [==============================] - 0s 6ms/step - loss: 1

Epoch 4/15
29/29 [==============================] - 0s 7ms/step - loss: 1.0543 - accuracy: 0.5908 - val_loss: 0.9642 - val_accuracy: 0.6201
Epoch 5/15
29/29 [==============================] - 0s 8ms/step - loss: 1.0112 - accuracy: 0.6083 - val_loss: 0.9412 - val_accuracy: 0.6157
Epoch 6/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9819 - accuracy: 0.6072 - val_loss: 0.9243 - val_accuracy: 0.6157
Epoch 7/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9575 - accuracy: 0.6280 - val_loss: 0.9076 - val_accuracy: 0.6245
Epoch 8/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9393 - accuracy: 0.6335 - val_loss: 0.9019 - val_accuracy: 0.6201
Epoch 9/15
29/29 [==============================] - 0s 6ms/step - loss: 0.9250 - accuracy: 0.6335 - val_loss: 0.8944 - val_accuracy: 0.6245
Epoch 10/15
29/29 [==============================] - 0s 6ms/step - loss: 0.9068 - accuracy: 0.6357 - val_loss: 0.8954 - val_accuracy: 0.6026
Epoch 11/15
29/29 [

8/8 [==============================] - 0s 2ms/step - loss: 0.8746 - accuracy: 0.6157
Epoch 1/15
29/29 [==============================] - 1s 12ms/step - loss: 2.3654 - accuracy: 0.1444 - val_loss: 2.1415 - val_accuracy: 0.3406
Epoch 2/15
29/29 [==============================] - 0s 4ms/step - loss: 1.9932 - accuracy: 0.4147 - val_loss: 1.8095 - val_accuracy: 0.4367
Epoch 3/15
29/29 [==============================] - 0s 5ms/step - loss: 1.7179 - accuracy: 0.4442 - val_loss: 1.5686 - val_accuracy: 0.4716
Epoch 4/15
29/29 [==============================] - 0s 4ms/step - loss: 1.5451 - accuracy: 0.4606 - val_loss: 1.4029 - val_accuracy: 0.4978
Epoch 5/15
29/29 [==============================] - 0s 10ms/step - loss: 1.4276 - accuracy: 0.4945 - val_loss: 1.2950 - val_accuracy: 0.5328
Epoch 6/15
29/29 [==============================] - 0s 5ms/step - loss: 1.3457 - accuracy: 0.5328 - val_loss: 1.2231 - val_accuracy: 0.5633
Epoch 7/15
29/29 [==============================] - 0s 6ms/step - loss: 1

Epoch 11/15
29/29 [==============================] - 0s 6ms/step - loss: 1.1058 - accuracy: 0.5733 - val_loss: 1.0047 - val_accuracy: 0.6288
Epoch 12/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0867 - accuracy: 0.5799 - val_loss: 0.9929 - val_accuracy: 0.6332
Epoch 13/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0870 - accuracy: 0.5788 - val_loss: 0.9852 - val_accuracy: 0.6201
Epoch 14/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0804 - accuracy: 0.5777 - val_loss: 0.9756 - val_accuracy: 0.6376
Epoch 15/15
8/8 [==============================] - 0s 2ms/step - loss: 0.9674 - accuracy: 0.6376
Epoch 1/15
29/29 [==============================] - 1s 12ms/step - loss: 1.7937 - accuracy: 0.4245 - val_loss: 1.3057 - val_accuracy: 0.5808
Epoch 2/15
29/29 [==============================] - 0s 4ms/step - loss: 1.2722 - accuracy: 0.5339 - val_loss: 1.0469 - val_accuracy: 0.6026
Epoch 3/15
29/29 [==============================] - 0s 4ms

Epoch 7/15
29/29 [==============================] - 0s 11ms/step - loss: 1.0135 - accuracy: 0.5875 - val_loss: 0.8981 - val_accuracy: 0.6332
Epoch 8/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9793 - accuracy: 0.6138 - val_loss: 0.8856 - val_accuracy: 0.6507
Epoch 9/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9887 - accuracy: 0.5996 - val_loss: 0.8804 - val_accuracy: 0.6550
Epoch 10/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9782 - accuracy: 0.6116 - val_loss: 0.8775 - val_accuracy: 0.6376
Epoch 11/15
29/29 [==============================] - 0s 4ms/step - loss: 0.9546 - accuracy: 0.6050 - val_loss: 0.8697 - val_accuracy: 0.6419
Epoch 12/15
29/29 [==============================] - 0s 4ms/step - loss: 0.9352 - accuracy: 0.6193 - val_loss: 0.8704 - val_accuracy: 0.6419
Epoch 13/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9226 - accuracy: 0.6236 - val_loss: 0.8650 - val_accuracy: 0.6550
Epoch 14/15
29/

Epoch 3/15
29/29 [==============================] - 0s 5ms/step - loss: 1.2234 - accuracy: 0.5350 - val_loss: 0.9883 - val_accuracy: 0.6070
Epoch 4/15
29/29 [==============================] - 0s 6ms/step - loss: 1.1714 - accuracy: 0.5536 - val_loss: 0.9485 - val_accuracy: 0.6026
Epoch 5/15
29/29 [==============================] - 0s 5ms/step - loss: 1.1196 - accuracy: 0.5810 - val_loss: 0.9375 - val_accuracy: 0.6070
Epoch 6/15
29/29 [==============================] - 0s 4ms/step - loss: 1.0662 - accuracy: 0.5689 - val_loss: 0.9216 - val_accuracy: 0.6114
Epoch 7/15
29/29 [==============================] - 0s 4ms/step - loss: 1.0601 - accuracy: 0.5832 - val_loss: 0.9069 - val_accuracy: 0.6288
Epoch 8/15
29/29 [==============================] - 0s 4ms/step - loss: 1.0149 - accuracy: 0.5941 - val_loss: 0.9004 - val_accuracy: 0.6201
Epoch 9/15
29/29 [==============================] - 0s 4ms/step - loss: 1.0016 - accuracy: 0.6072 - val_loss: 0.8909 - val_accuracy: 0.6376
Epoch 10/15
29/29 [=

29/29 [==============================] - 0s 4ms/step - loss: 0.8069 - accuracy: 0.6674 - val_loss: 0.9033 - val_accuracy: 0.6245
Epoch 15/15
8/8 [==============================] - 0s 2ms/step - loss: 0.9067 - accuracy: 0.6070
Epoch 1/15
29/29 [==============================] - 1s 15ms/step - loss: 2.2078 - accuracy: 0.3370 - val_loss: 2.0073 - val_accuracy: 0.4323
Epoch 2/15
29/29 [==============================] - 0s 5ms/step - loss: 1.8784 - accuracy: 0.4508 - val_loss: 1.7004 - val_accuracy: 0.4847
Epoch 3/15
29/29 [==============================] - 0s 4ms/step - loss: 1.6222 - accuracy: 0.4891 - val_loss: 1.4803 - val_accuracy: 0.5590
Epoch 4/15
29/29 [==============================] - 0s 4ms/step - loss: 1.4487 - accuracy: 0.5186 - val_loss: 1.3391 - val_accuracy: 0.5633
Epoch 5/15
29/29 [==============================] - 0s 4ms/step - loss: 1.3369 - accuracy: 0.5263 - val_loss: 1.2467 - val_accuracy: 0.5721
Epoch 6/15
29/29 [==============================] - 0s 4ms/step - loss: 1

Epoch 10/15
29/29 [==============================] - 0s 4ms/step - loss: 1.1308 - accuracy: 0.5755 - val_loss: 1.0507 - val_accuracy: 0.5939
Epoch 11/15
29/29 [==============================] - 0s 5ms/step - loss: 1.1164 - accuracy: 0.5799 - val_loss: 1.0304 - val_accuracy: 0.5895
Epoch 12/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0997 - accuracy: 0.5744 - val_loss: 1.0163 - val_accuracy: 0.5939
Epoch 13/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0881 - accuracy: 0.5875 - val_loss: 1.0034 - val_accuracy: 0.6114
Epoch 14/15
29/29 [==============================] - 0s 9ms/step - loss: 1.0690 - accuracy: 0.5897 - val_loss: 0.9934 - val_accuracy: 0.6114
Epoch 15/15
8/8 [==============================] - 0s 3ms/step - loss: 0.9837 - accuracy: 0.6114
Epoch 1/15
29/29 [==============================] - 1s 17ms/step - loss: 1.5463 - accuracy: 0.5109 - val_loss: 1.1036 - val_accuracy: 0.5764
Epoch 2/15
29/29 [==============================] - 0s 12

Epoch 6/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9610 - accuracy: 0.6105 - val_loss: 0.8927 - val_accuracy: 0.6419
Epoch 7/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9234 - accuracy: 0.6477 - val_loss: 0.8945 - val_accuracy: 0.6332
Epoch 8/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9087 - accuracy: 0.6400 - val_loss: 0.8893 - val_accuracy: 0.6114
Epoch 9/15
29/29 [==============================] - 0s 5ms/step - loss: 0.8990 - accuracy: 0.6346 - val_loss: 0.8956 - val_accuracy: 0.5983
Epoch 10/15
29/29 [==============================] - 0s 4ms/step - loss: 0.8838 - accuracy: 0.6247 - val_loss: 0.8899 - val_accuracy: 0.6201
Epoch 11/15
29/29 [==============================] - 0s 5ms/step - loss: 0.8786 - accuracy: 0.6499 - val_loss: 0.8781 - val_accuracy: 0.6245
Epoch 12/15
29/29 [==============================] - 0s 5ms/step - loss: 0.8541 - accuracy: 0.6521 - val_loss: 0.8837 - val_accuracy: 0.6594
Epoch 13/15
29/29

Epoch 2/15
29/29 [==============================] - 0s 4ms/step - loss: 1.1762 - accuracy: 0.5438 - val_loss: 0.9772 - val_accuracy: 0.5983
Epoch 3/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0772 - accuracy: 0.5656 - val_loss: 0.9273 - val_accuracy: 0.6114
Epoch 4/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0085 - accuracy: 0.5996 - val_loss: 0.9085 - val_accuracy: 0.6245
Epoch 5/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9936 - accuracy: 0.5974 - val_loss: 0.8901 - val_accuracy: 0.6245
Epoch 6/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9810 - accuracy: 0.6061 - val_loss: 0.8952 - val_accuracy: 0.6070
Epoch 7/15
29/29 [==============================] - 0s 4ms/step - loss: 0.9435 - accuracy: 0.6160 - val_loss: 0.8820 - val_accuracy: 0.6026
Epoch 8/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9359 - accuracy: 0.6225 - val_loss: 0.8946 - val_accuracy: 0.6070
Epoch 9/15
29/29 [==

29/29 [==============================] - 0s 5ms/step - loss: 0.9128 - accuracy: 0.6335 - val_loss: 0.8665 - val_accuracy: 0.6201
Epoch 14/15
29/29 [==============================] - 0s 5ms/step - loss: 0.8966 - accuracy: 0.6433 - val_loss: 0.8739 - val_accuracy: 0.6332
Epoch 15/15
8/8 [==============================] - 0s 3ms/step - loss: 0.8719 - accuracy: 0.6201
Epoch 1/15
29/29 [==============================] - 1s 15ms/step - loss: 2.1657 - accuracy: 0.2713 - val_loss: 1.9272 - val_accuracy: 0.4410
Epoch 2/15
29/29 [==============================] - 0s 6ms/step - loss: 1.8628 - accuracy: 0.4136 - val_loss: 1.6425 - val_accuracy: 0.5153
Epoch 3/15
29/29 [==============================] - 0s 5ms/step - loss: 1.6326 - accuracy: 0.4737 - val_loss: 1.4418 - val_accuracy: 0.5502
Epoch 4/15
29/29 [==============================] - 0s 5ms/step - loss: 1.4933 - accuracy: 0.4880 - val_loss: 1.3143 - val_accuracy: 0.5371
Epoch 5/15
29/29 [==============================] - 0s 4ms/step - loss: 

Epoch 9/15
29/29 [==============================] - 0s 4ms/step - loss: 1.1200 - accuracy: 0.5842 - val_loss: 1.0307 - val_accuracy: 0.6070
Epoch 10/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0961 - accuracy: 0.5821 - val_loss: 1.0109 - val_accuracy: 0.5983
Epoch 11/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0784 - accuracy: 0.5864 - val_loss: 0.9962 - val_accuracy: 0.6026
Epoch 12/15
29/29 [==============================] - 0s 4ms/step - loss: 1.0634 - accuracy: 0.6083 - val_loss: 0.9824 - val_accuracy: 0.6070
Epoch 13/15
29/29 [==============================] - 0s 13ms/step - loss: 1.0513 - accuracy: 0.6050 - val_loss: 0.9716 - val_accuracy: 0.6419
Epoch 14/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0397 - accuracy: 0.6149 - val_loss: 0.9670 - val_accuracy: 0.6114
Epoch 15/15
8/8 [==============================] - 0s 3ms/step - loss: 0.9585 - accuracy: 0.6288
Epoch 1/15
29/29 [==============================] - 1s 20

Epoch 5/15
29/29 [==============================] - 0s 6ms/step - loss: 0.9218 - accuracy: 0.6258 - val_loss: 0.8912 - val_accuracy: 0.6419
Epoch 6/15
29/29 [==============================] - 0s 7ms/step - loss: 0.8892 - accuracy: 0.6488 - val_loss: 0.8991 - val_accuracy: 0.5939
Epoch 7/15
29/29 [==============================] - 0s 7ms/step - loss: 0.8768 - accuracy: 0.6510 - val_loss: 0.8962 - val_accuracy: 0.6114
Epoch 8/15
29/29 [==============================] - 0s 6ms/step - loss: 0.8554 - accuracy: 0.6597 - val_loss: 0.9301 - val_accuracy: 0.6070
Epoch 9/15
29/29 [==============================] - 0s 7ms/step - loss: 0.8445 - accuracy: 0.6619 - val_loss: 0.8876 - val_accuracy: 0.6463
Epoch 10/15
29/29 [==============================] - 0s 7ms/step - loss: 0.8264 - accuracy: 0.6696 - val_loss: 0.8946 - val_accuracy: 0.6070
Epoch 11/15
29/29 [==============================] - 0s 7ms/step - loss: 0.8125 - accuracy: 0.6575 - val_loss: 0.8754 - val_accuracy: 0.6201
Epoch 12/15
29/29 

8/8 [==============================] - 0s 4ms/step - loss: 0.9097 - accuracy: 0.6157
Epoch 1/15
29/29 [==============================] - 3s 12ms/step - loss: 1.4209 - accuracy: 0.5394 - val_loss: 0.9989 - val_accuracy: 0.5895
Epoch 2/15
29/29 [==============================] - 0s 5ms/step - loss: 1.0485 - accuracy: 0.5810 - val_loss: 0.9151 - val_accuracy: 0.6070
Epoch 3/15
29/29 [==============================] - 0s 6ms/step - loss: 0.9880 - accuracy: 0.6083 - val_loss: 0.8974 - val_accuracy: 0.6070
Epoch 4/15
29/29 [==============================] - 0s 6ms/step - loss: 0.9562 - accuracy: 0.6039 - val_loss: 0.8754 - val_accuracy: 0.6419
Epoch 5/15
29/29 [==============================] - 0s 5ms/step - loss: 0.9410 - accuracy: 0.6258 - val_loss: 0.8944 - val_accuracy: 0.6026
Epoch 6/15
29/29 [==============================] - 0s 6ms/step - loss: 0.9229 - accuracy: 0.6280 - val_loss: 0.8895 - val_accuracy: 0.6070
Epoch 7/15
29/29 [==============================] - 0s 6ms/step - loss: 0.

Epoch 12/15
29/29 [==============================] - 0s 5ms/step - loss: 0.8327 - accuracy: 0.6499 - val_loss: 0.8959 - val_accuracy: 0.6245
Epoch 13/15
29/29 [==============================] - 0s 5ms/step - loss: 0.8199 - accuracy: 0.6816 - val_loss: 0.9171 - val_accuracy: 0.6114
Epoch 14/15
29/29 [==============================] - 0s 5ms/step - loss: 0.7972 - accuracy: 0.6663 - val_loss: 0.8818 - val_accuracy: 0.5983
Epoch 15/15
8/8 [==============================] - 0s 3ms/step - loss: 0.9146 - accuracy: 0.6332
Epoch 1/15
29/29 [==============================] - 1s 13ms/step - loss: 2.1731 - accuracy: 0.3162 - val_loss: 1.9420 - val_accuracy: 0.4279
Epoch 2/15
29/29 [==============================] - 0s 6ms/step - loss: 1.8311 - accuracy: 0.4628 - val_loss: 1.6373 - val_accuracy: 0.4978
Epoch 3/15
29/29 [==============================] - 0s 5ms/step - loss: 1.5887 - accuracy: 0.5077 - val_loss: 1.4325 - val_accuracy: 0.5240
Epoch 4/15
29/29 [==============================] - 0s 6ms/

Epoch 8/15
29/29 [==============================] - 0s 5ms/step - loss: 1.1741 - accuracy: 0.5613 - val_loss: 1.0562 - val_accuracy: 0.5983
Epoch 9/15
29/29 [==============================] - 0s 5ms/step - loss: 1.1605 - accuracy: 0.5624 - val_loss: 1.0337 - val_accuracy: 0.6070
Epoch 10/15
29/29 [==============================] - 0s 6ms/step - loss: 1.1470 - accuracy: 0.5547 - val_loss: 1.0185 - val_accuracy: 0.6070
Epoch 11/15
29/29 [==============================] - 0s 6ms/step - loss: 1.1283 - accuracy: 0.5613 - val_loss: 1.0005 - val_accuracy: 0.6070
Epoch 12/15
29/29 [==============================] - 0s 6ms/step - loss: 1.1095 - accuracy: 0.5897 - val_loss: 0.9872 - val_accuracy: 0.6114
Epoch 13/15
29/29 [==============================] - 0s 7ms/step - loss: 1.1010 - accuracy: 0.5875 - val_loss: 0.9807 - val_accuracy: 0.6201
Epoch 14/15
29/29 [==============================] - 0s 6ms/step - loss: 1.0725 - accuracy: 0.5886 - val_loss: 0.9722 - val_accuracy: 0.6201
Epoch 15/15
8/8

29/29 [==============================] - 0s 5ms/step - loss: 1.0721 - accuracy: 0.5930 - val_loss: 1.0027 - val_accuracy: 0.5677
Epoch 15/20
29/29 [==============================] - 0s 6ms/step - loss: 1.0604 - accuracy: 0.5963 - val_loss: 0.9909 - val_accuracy: 0.5721
Epoch 16/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0504 - accuracy: 0.6007 - val_loss: 0.9835 - val_accuracy: 0.5808
Epoch 17/20
29/29 [==============================] - 0s 6ms/step - loss: 1.0418 - accuracy: 0.6105 - val_loss: 0.9750 - val_accuracy: 0.5895
Epoch 18/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0336 - accuracy: 0.6061 - val_loss: 0.9692 - val_accuracy: 0.6114
Epoch 19/20
29/29 [==============================] - 0s 6ms/step - loss: 1.0263 - accuracy: 0.6105 - val_loss: 0.9626 - val_accuracy: 0.6157
Epoch 20/20
8/8 [==============================] - 0s 3ms/step - loss: 0.9589 - accuracy: 0.6157
Epoch 1/20
29/29 [==============================] - 1s 14ms/step - lo

29/29 [==============================] - 0s 5ms/step - loss: 1.1784 - accuracy: 0.5470 - val_loss: 1.0813 - val_accuracy: 0.5677
Epoch 11/20
29/29 [==============================] - 0s 4ms/step - loss: 1.1430 - accuracy: 0.5711 - val_loss: 1.0555 - val_accuracy: 0.5764
Epoch 12/20
29/29 [==============================] - 0s 8ms/step - loss: 1.1285 - accuracy: 0.5635 - val_loss: 1.0352 - val_accuracy: 0.5852
Epoch 13/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0982 - accuracy: 0.5744 - val_loss: 1.0179 - val_accuracy: 0.5808
Epoch 14/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0903 - accuracy: 0.5656 - val_loss: 1.0031 - val_accuracy: 0.5808
Epoch 15/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0859 - accuracy: 0.5646 - val_loss: 0.9923 - val_accuracy: 0.5895
Epoch 16/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0585 - accuracy: 0.5821 - val_loss: 0.9821 - val_accuracy: 0.5983
Epoch 17/20
29/29 [======

29/29 [==============================] - 0s 4ms/step - loss: 1.3356 - accuracy: 0.5055 - val_loss: 1.2255 - val_accuracy: 0.5328
Epoch 7/20
29/29 [==============================] - 0s 4ms/step - loss: 1.2819 - accuracy: 0.5186 - val_loss: 1.1821 - val_accuracy: 0.5502
Epoch 8/20
29/29 [==============================] - 0s 5ms/step - loss: 1.2243 - accuracy: 0.5328 - val_loss: 1.1506 - val_accuracy: 0.5459
Epoch 9/20
29/29 [==============================] - 0s 6ms/step - loss: 1.2075 - accuracy: 0.5197 - val_loss: 1.1232 - val_accuracy: 0.5546
Epoch 10/20
29/29 [==============================] - 0s 5ms/step - loss: 1.1800 - accuracy: 0.5438 - val_loss: 1.1035 - val_accuracy: 0.5502
Epoch 11/20
29/29 [==============================] - 0s 6ms/step - loss: 1.1730 - accuracy: 0.5350 - val_loss: 1.0850 - val_accuracy: 0.5633
Epoch 12/20
29/29 [==============================] - 0s 5ms/step - loss: 1.1591 - accuracy: 0.5481 - val_loss: 1.0689 - val_accuracy: 0.5677
Epoch 13/20
29/29 [=========

29/29 [==============================] - 0s 5ms/step - loss: 1.9633 - accuracy: 0.4300 - val_loss: 1.7700 - val_accuracy: 0.4672
Epoch 3/20
29/29 [==============================] - 0s 5ms/step - loss: 1.7050 - accuracy: 0.5000 - val_loss: 1.5327 - val_accuracy: 0.4716
Epoch 4/20
29/29 [==============================] - 0s 5ms/step - loss: 1.5152 - accuracy: 0.5088 - val_loss: 1.3757 - val_accuracy: 0.4760
Epoch 5/20
29/29 [==============================] - 0s 4ms/step - loss: 1.3904 - accuracy: 0.5394 - val_loss: 1.2724 - val_accuracy: 0.5153
Epoch 6/20
29/29 [==============================] - 0s 5ms/step - loss: 1.3399 - accuracy: 0.5328 - val_loss: 1.1997 - val_accuracy: 0.5502
Epoch 7/20
29/29 [==============================] - 0s 5ms/step - loss: 1.2849 - accuracy: 0.5383 - val_loss: 1.1486 - val_accuracy: 0.5590
Epoch 8/20
29/29 [==============================] - 0s 5ms/step - loss: 1.2391 - accuracy: 0.5427 - val_loss: 1.1112 - val_accuracy: 0.5677
Epoch 9/20
29/29 [=============

29/29 [==============================] - 0s 4ms/step - loss: 0.9008 - accuracy: 0.6357 - val_loss: 0.8705 - val_accuracy: 0.6114
Epoch 20/20
8/8 [==============================] - 0s 2ms/step - loss: 0.8623 - accuracy: 0.6157
Epoch 1/20
29/29 [==============================] - 1s 13ms/step - loss: 2.3752 - accuracy: 0.1247 - val_loss: 2.1562 - val_accuracy: 0.3799
Epoch 2/20
29/29 [==============================] - 0s 5ms/step - loss: 2.0567 - accuracy: 0.3840 - val_loss: 1.8703 - val_accuracy: 0.4847
Epoch 3/20
29/29 [==============================] - 0s 5ms/step - loss: 1.8174 - accuracy: 0.4409 - val_loss: 1.6406 - val_accuracy: 0.4891
Epoch 4/20
29/29 [==============================] - 0s 5ms/step - loss: 1.6295 - accuracy: 0.4814 - val_loss: 1.4642 - val_accuracy: 0.5109
Epoch 5/20
29/29 [==============================] - 0s 5ms/step - loss: 1.4964 - accuracy: 0.5142 - val_loss: 1.3348 - val_accuracy: 0.5197
Epoch 6/20
29/29 [==============================] - 0s 5ms/step - loss: 1

29/29 [==============================] - 0s 5ms/step - loss: 0.9594 - accuracy: 0.6138 - val_loss: 0.8802 - val_accuracy: 0.6201
Epoch 16/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9275 - accuracy: 0.6204 - val_loss: 0.8819 - val_accuracy: 0.6114
Epoch 17/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9505 - accuracy: 0.6149 - val_loss: 0.8821 - val_accuracy: 0.6026
Epoch 18/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9352 - accuracy: 0.6313 - val_loss: 0.8763 - val_accuracy: 0.6026
Epoch 19/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9282 - accuracy: 0.6236 - val_loss: 0.8769 - val_accuracy: 0.5983
Epoch 20/20
8/8 [==============================] - 0s 2ms/step - loss: 0.8765 - accuracy: 0.5983
Epoch 1/20
29/29 [==============================] - 1s 13ms/step - loss: 2.3543 - accuracy: 0.1860 - val_loss: 2.0234 - val_accuracy: 0.3755
Epoch 2/20
29/29 [==============================] - 0s 11ms/step - lo

29/29 [==============================] - 0s 4ms/step - loss: 0.8438 - accuracy: 0.6488 - val_loss: 0.8950 - val_accuracy: 0.6201
Epoch 12/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8260 - accuracy: 0.6652 - val_loss: 0.8832 - val_accuracy: 0.6376
Epoch 13/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8217 - accuracy: 0.6641 - val_loss: 0.8719 - val_accuracy: 0.6332
Epoch 14/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8081 - accuracy: 0.6805 - val_loss: 0.8872 - val_accuracy: 0.6201
Epoch 15/20
29/29 [==============================] - 0s 3ms/step - loss: 0.7888 - accuracy: 0.6751 - val_loss: 0.8836 - val_accuracy: 0.6332
Epoch 16/20
29/29 [==============================] - 0s 4ms/step - loss: 0.7814 - accuracy: 0.6630 - val_loss: 0.8899 - val_accuracy: 0.6201
Epoch 17/20
29/29 [==============================] - 0s 5ms/step - loss: 0.7828 - accuracy: 0.6685 - val_loss: 0.9008 - val_accuracy: 0.6026
Epoch 18/20
29/29 [======

29/29 [==============================] - 0s 4ms/step - loss: 0.9101 - accuracy: 0.6357 - val_loss: 0.8825 - val_accuracy: 0.6594
Epoch 8/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8889 - accuracy: 0.6565 - val_loss: 0.9107 - val_accuracy: 0.5983
Epoch 9/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8961 - accuracy: 0.6389 - val_loss: 0.9008 - val_accuracy: 0.5939
Epoch 10/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8753 - accuracy: 0.6499 - val_loss: 0.8905 - val_accuracy: 0.6157
Epoch 11/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8595 - accuracy: 0.6455 - val_loss: 0.8764 - val_accuracy: 0.5983
Epoch 12/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8476 - accuracy: 0.6488 - val_loss: 0.8810 - val_accuracy: 0.5939
Epoch 13/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8496 - accuracy: 0.6379 - val_loss: 0.8888 - val_accuracy: 0.6114
Epoch 14/20
29/29 [========

29/29 [==============================] - 0s 6ms/step - loss: 1.0630 - accuracy: 0.5810 - val_loss: 0.9364 - val_accuracy: 0.6376
Epoch 4/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9940 - accuracy: 0.6083 - val_loss: 0.9312 - val_accuracy: 0.6157
Epoch 5/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9759 - accuracy: 0.6193 - val_loss: 0.9199 - val_accuracy: 0.6376
Epoch 6/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9490 - accuracy: 0.6346 - val_loss: 0.8956 - val_accuracy: 0.6026
Epoch 7/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9395 - accuracy: 0.6105 - val_loss: 0.8980 - val_accuracy: 0.6245
Epoch 8/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9124 - accuracy: 0.6368 - val_loss: 0.8892 - val_accuracy: 0.6332
Epoch 9/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9065 - accuracy: 0.6335 - val_loss: 0.8969 - val_accuracy: 0.6245
Epoch 10/20
29/29 [============

8/8 [==============================] - 0s 3ms/step - loss: 0.9096 - accuracy: 0.6332
Epoch 1/20
29/29 [==============================] - 2s 12ms/step - loss: 1.7626 - accuracy: 0.4617 - val_loss: 1.1871 - val_accuracy: 0.5328
Epoch 2/20
29/29 [==============================] - 0s 5ms/step - loss: 1.1609 - accuracy: 0.5635 - val_loss: 0.9894 - val_accuracy: 0.5983
Epoch 3/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0589 - accuracy: 0.5788 - val_loss: 0.9356 - val_accuracy: 0.6288
Epoch 4/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0173 - accuracy: 0.5974 - val_loss: 0.9363 - val_accuracy: 0.6245
Epoch 5/20
29/29 [==============================] - 0s 6ms/step - loss: 0.9850 - accuracy: 0.6018 - val_loss: 0.9141 - val_accuracy: 0.6245
Epoch 6/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9786 - accuracy: 0.6039 - val_loss: 0.9077 - val_accuracy: 0.6201
Epoch 7/20
29/29 [==============================] - 0s 5ms/step - loss: 0.

29/29 [==============================] - 0s 4ms/step - loss: 0.8362 - accuracy: 0.6685 - val_loss: 0.8737 - val_accuracy: 0.6157
Epoch 17/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8378 - accuracy: 0.6357 - val_loss: 0.8852 - val_accuracy: 0.6114
Epoch 18/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8311 - accuracy: 0.6608 - val_loss: 0.8804 - val_accuracy: 0.6463
Epoch 19/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8253 - accuracy: 0.6586 - val_loss: 0.8725 - val_accuracy: 0.5939
Epoch 20/20
8/8 [==============================] - 0s 3ms/step - loss: 0.8928 - accuracy: 0.6026
Epoch 1/20
29/29 [==============================] - 2s 15ms/step - loss: 1.7578 - accuracy: 0.4311 - val_loss: 1.1497 - val_accuracy: 0.6070
Epoch 2/20
29/29 [==============================] - 0s 5ms/step - loss: 1.1748 - accuracy: 0.5492 - val_loss: 0.9764 - val_accuracy: 0.6114
Epoch 3/20
29/29 [==============================] - 0s 4ms/step - loss

29/29 [==============================] - 0s 5ms/step - loss: 0.8865 - accuracy: 0.6422 - val_loss: 0.8937 - val_accuracy: 0.5852
Epoch 13/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8812 - accuracy: 0.6411 - val_loss: 0.8965 - val_accuracy: 0.6070
Epoch 14/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8752 - accuracy: 0.6521 - val_loss: 0.8820 - val_accuracy: 0.6288
Epoch 15/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8885 - accuracy: 0.6488 - val_loss: 0.8874 - val_accuracy: 0.6288
Epoch 16/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8648 - accuracy: 0.6674 - val_loss: 0.8892 - val_accuracy: 0.5764
Epoch 17/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8545 - accuracy: 0.6565 - val_loss: 0.8732 - val_accuracy: 0.6201
Epoch 18/20
29/29 [==============================] - 0s 4ms/step - loss: 0.8454 - accuracy: 0.6488 - val_loss: 0.8828 - val_accuracy: 0.6157
Epoch 19/20
29/29 [======

29/29 [==============================] - 0s 4ms/step - loss: 0.9802 - accuracy: 0.6116 - val_loss: 0.8750 - val_accuracy: 0.6288
Epoch 9/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9528 - accuracy: 0.6302 - val_loss: 0.8843 - val_accuracy: 0.6114
Epoch 10/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9311 - accuracy: 0.6269 - val_loss: 0.8844 - val_accuracy: 0.6114
Epoch 11/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9295 - accuracy: 0.6094 - val_loss: 0.8770 - val_accuracy: 0.6332
Epoch 12/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9024 - accuracy: 0.6335 - val_loss: 0.8789 - val_accuracy: 0.6288
Epoch 13/20
29/29 [==============================] - 0s 4ms/step - loss: 0.9128 - accuracy: 0.6160 - val_loss: 0.8897 - val_accuracy: 0.6114
Epoch 14/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8999 - accuracy: 0.6258 - val_loss: 0.8852 - val_accuracy: 0.6157
Epoch 15/20
29/29 [=======

29/29 [==============================] - 0s 5ms/step - loss: 0.9245 - accuracy: 0.6247 - val_loss: 0.8798 - val_accuracy: 0.6419
Epoch 5/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9030 - accuracy: 0.6204 - val_loss: 0.9290 - val_accuracy: 0.5939
Epoch 6/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8863 - accuracy: 0.6313 - val_loss: 0.8764 - val_accuracy: 0.6288
Epoch 7/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8598 - accuracy: 0.6543 - val_loss: 0.8812 - val_accuracy: 0.6201
Epoch 8/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8518 - accuracy: 0.6488 - val_loss: 0.8968 - val_accuracy: 0.6332
Epoch 9/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8289 - accuracy: 0.6685 - val_loss: 0.8821 - val_accuracy: 0.6026
Epoch 10/20
29/29 [==============================] - 0s 5ms/step - loss: 0.8195 - accuracy: 0.6586 - val_loss: 0.8792 - val_accuracy: 0.6332
Epoch 11/20
29/29 [===========

8/8 [==============================] - 0s 3ms/step - loss: 0.9266 - accuracy: 0.6332
Epoch 1/20
29/29 [==============================] - 1s 13ms/step - loss: 1.3189 - accuracy: 0.5284 - val_loss: 0.9347 - val_accuracy: 0.6507
Epoch 2/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0230 - accuracy: 0.6083 - val_loss: 0.8912 - val_accuracy: 0.6419
Epoch 3/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9767 - accuracy: 0.6083 - val_loss: 0.8928 - val_accuracy: 0.6026
Epoch 4/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9314 - accuracy: 0.6182 - val_loss: 0.8797 - val_accuracy: 0.6376
Epoch 5/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9168 - accuracy: 0.6127 - val_loss: 0.8939 - val_accuracy: 0.6114
Epoch 6/20
29/29 [==============================] - 0s 6ms/step - loss: 0.8846 - accuracy: 0.6258 - val_loss: 0.8730 - val_accuracy: 0.6201
Epoch 7/20
29/29 [==============================] - 0s 5ms/step - loss: 0.

29/29 [==============================] - 0s 5ms/step - loss: 1.0225 - accuracy: 0.6094 - val_loss: 0.9317 - val_accuracy: 0.6157
Epoch 18/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0122 - accuracy: 0.6028 - val_loss: 0.9255 - val_accuracy: 0.6550
Epoch 19/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0048 - accuracy: 0.6116 - val_loss: 0.9223 - val_accuracy: 0.6463
Epoch 20/20
8/8 [==============================] - 0s 3ms/step - loss: 0.9166 - accuracy: 0.6594
Epoch 1/20
29/29 [==============================] - 1s 13ms/step - loss: 1.2801 - accuracy: 0.5295 - val_loss: 0.9593 - val_accuracy: 0.6157
Epoch 2/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0142 - accuracy: 0.6007 - val_loss: 0.9438 - val_accuracy: 0.5939
Epoch 3/20
29/29 [==============================] - 0s 5ms/step - loss: 0.9793 - accuracy: 0.6039 - val_loss: 0.9007 - val_accuracy: 0.6114
Epoch 4/20
29/29 [==============================] - 0s 5ms/step - loss:

29/29 [==============================] - 0s 5ms/step - loss: 1.0654 - accuracy: 0.5919 - val_loss: 0.9743 - val_accuracy: 0.6114
Epoch 14/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0628 - accuracy: 0.5853 - val_loss: 0.9643 - val_accuracy: 0.6114
Epoch 15/20
29/29 [==============================] - 0s 4ms/step - loss: 1.0490 - accuracy: 0.5952 - val_loss: 0.9564 - val_accuracy: 0.6070
Epoch 16/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0372 - accuracy: 0.6018 - val_loss: 0.9506 - val_accuracy: 0.6157
Epoch 17/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0347 - accuracy: 0.6007 - val_loss: 0.9452 - val_accuracy: 0.6070
Epoch 18/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0221 - accuracy: 0.6072 - val_loss: 0.9380 - val_accuracy: 0.6201
Epoch 19/20
29/29 [==============================] - 0s 4ms/step - loss: 1.0279 - accuracy: 0.6050 - val_loss: 0.9366 - val_accuracy: 0.6026
Epoch 20/20
8/8 [========

29/29 [==============================] - 0s 4ms/step - loss: 1.1349 - accuracy: 0.5700 - val_loss: 1.0329 - val_accuracy: 0.6070
Epoch 10/20
29/29 [==============================] - 0s 5ms/step - loss: 1.1345 - accuracy: 0.5624 - val_loss: 1.0135 - val_accuracy: 0.6114
Epoch 11/20
29/29 [==============================] - 0s 4ms/step - loss: 1.1010 - accuracy: 0.5744 - val_loss: 0.9957 - val_accuracy: 0.6114
Epoch 12/20
29/29 [==============================] - 0s 4ms/step - loss: 1.0920 - accuracy: 0.5755 - val_loss: 0.9829 - val_accuracy: 0.6114
Epoch 13/20
29/29 [==============================] - 0s 4ms/step - loss: 1.0703 - accuracy: 0.5864 - val_loss: 0.9705 - val_accuracy: 0.6114
Epoch 14/20
29/29 [==============================] - 0s 5ms/step - loss: 1.0717 - accuracy: 0.5799 - val_loss: 0.9629 - val_accuracy: 0.6245
Epoch 15/20
29/29 [==============================] - 0s 4ms/step - loss: 1.0579 - accuracy: 0.5799 - val_loss: 0.9548 - val_accuracy: 0.6201
Epoch 16/20
29/29 [======

29/29 [==============================] - 0s 5ms/step - loss: 1.3610 - accuracy: 0.5470 - val_loss: 1.2433 - val_accuracy: 0.5939
Epoch 6/20
29/29 [==============================] - 0s 5ms/step - loss: 1.2961 - accuracy: 0.5142 - val_loss: 1.1754 - val_accuracy: 0.5852
Epoch 7/20
29/29 [==============================] - 0s 7ms/step - loss: 1.2455 - accuracy: 0.5470 - val_loss: 1.1271 - val_accuracy: 0.5939
Epoch 8/20
29/29 [==============================] - 0s 6ms/step - loss: 1.1935 - accuracy: 0.5569 - val_loss: 1.0902 - val_accuracy: 0.5895
Epoch 9/20
29/29 [==============================] - 0s 5ms/step - loss: 1.1749 - accuracy: 0.5438 - val_loss: 1.0624 - val_accuracy: 0.5939
Epoch 10/20
29/29 [==============================] - 0s 6ms/step - loss: 1.1310 - accuracy: 0.5821 - val_loss: 1.0403 - val_accuracy: 0.5983
Epoch 11/20
29/29 [==============================] - 0s 8ms/step - loss: 1.1218 - accuracy: 0.5722 - val_loss: 1.0234 - val_accuracy: 0.5983
Epoch 12/20
29/29 [==========

In [382]:
end_time = datetime.datetime.now()

In [383]:
run_time = end_time - start_time
run_time.total_seconds()
seconds_total = run_time.total_seconds()
hours   = divmod(seconds_total, 3600)
minutes = divmod(hours[1], 60)
seconds = divmod(minutes[1], 1)
print("Total runtime with itteration thru the parameters: %d hours, %d minutes and %d seconds" % (hours[0], minutes[0], seconds[0]))

Total runtime with itteration thru the parameters: 0 hours, 12 minutes and 5 seconds


In [384]:
test_results.sort_values(by='accuracy', ascending=False).head()

epochs  neurons  dropout Optimazion type  accuracy
32        5      128      0.4         rmsprop  0.659389
205      20      256      0.2             sgd  0.659389
153      15      256      0.3            adam  0.655022
122      15       64      0.4         rmsprop  0.655022
101      10      256      0.3         rmsprop  0.655022

In [385]:
test_results.sort_values(by='accuracy', ascending=False).tail()

epochs  neurons  dropout Optimazion type  accuracy
19       5      128      0.0             sgd  0.541485
25       5      128      0.2             sgd  0.541485
10       5       64      0.3             sgd  0.537118
28       5      128      0.3             sgd  0.537118
16       5       64      0.5             sgd  0.497817

In [386]:
test_results.to_csv('My_iterations/data_'+ start_time.strftime("%Y%m%d") +'_starttime' + start_time.strftime("%H%M") + '_duration' + str(hours[0]) + str(minutes[0]) + '.csv', index=False)  

### Delkonklusion ###  

#### Tensorflow Board ####  
  
  Tensorboardet gir et godt overblik over epochs og loss. Det bliver tydeligt hvornår, både epochs og loss begynder og flade ud og hvor der evt. kunne være brug for et early stop.  
  
Tensorboardet gir også et indtryk, i graferne for epochs vs. iteration og loss vs. iteration, af over og under fitting.  

#### Gennemløb ####
  
Ud fra tre gennemløb af vores test af modellen kan vi også i denne version, konkludere følgende med nogenlunde sikkerhed:  
  * Det betyder rigtig meget for modellens accuracy, hvordan den er sat sammen (vi får i dette tilfælde resultater fra ca. 49% til ca 66%).  
  * Vi kan efter de tre gennemløb se at vi overvejende får et godt resultat ved at bruge rmsprop optimeringsmetoden. Vi kan anvende sgd og adam, men det kræver et stort antal neuroner og mange epochs.  
  * Vi skal undgå at bruge sgd optimeringsmetoden sammen med lavt epochs og lavt neuron tal, da dette vil resultere i lav score.  
  
Vi svært ved at sige noget om:
  
  * Når vi anvender rmsprop som optimering, er det sværet at finde mønster i epochs og dropout.

Det vi kan sige er at det ser ud til at være mest fornuftigt at bruge rmsprop når man som i vores tilfælde bruger ANN og kun et lag. Det ser også, ikke uventet, ud til at et stort antal neuroner samt et rimeligt antal dropouts gir det bedste resultat.  
  

**Figur 1: Oversigt over tre bedste placeringer i tre gennemløb.**  
<img align="left" src="best.png"><br><br><br><br><br><br><br><br><br><br>
  
  
**Figur 2: Oversigt over tre værste placeringer i tre gennemløb.**  
<img align="left" src="worst.png"> 